File: csharp/sdk/src/ModelContextProtocol.Interceptors/Gateway/GatewayInterceptorClientProvider.cs
using ModelContextProtocol.Client;
using ModelContextProtocol.Server;

namespace ModelContextProtocol.Interceptors.Gateway;

internal sealed class GatewayInterceptorClientProvider : IAsyncDisposable
{
    private readonly IReadOnlyList<McpClient> _staticClients;
    private readonly Func<MessageContext, string, CancellationToken, ValueTask<IReadOnlyList<McpInterceptorServerConnectionOptions>>>? _connectionResolver;
    private readonly GatewayInterceptorClientPool? _clientPool;

    internal GatewayInterceptorClientProvider(
        IReadOnlyList<McpClient> staticClients,
        Func<MessageContext, string, CancellationToken, ValueTask<IReadOnlyList<McpInterceptorServerConnectionOptions>>>? connectionResolver)
    {
        _staticClients = staticClients;
        _connectionResolver = connectionResolver;
        _clientPool = connectionResolver is null ? null : new GatewayInterceptorClientPool();
    }

    internal IReadOnlyList<McpClient> StaticClients => _staticClients;

    internal async ValueTask<GatewayResolvedInterceptorClients> ResolveAsync(
        MessageContext messageContext,
        string @event,
        CancellationToken cancellationToken)
    {
        if (_connectionResolver is null)
        {
            return new GatewayResolvedInterceptorClients(_staticClients);
        }

        var resolvedConnections = await _connectionResolver(messageContext, @event, cancellationToken) ?? [];
        var resolvedDynamicClients = await _clientPool!.ResolveClientsAsync(resolvedConnections, cancellationToken);

        if (_staticClients.Count == 0)
        {
            return resolvedDynamicClients;
        }

        if (resolvedDynamicClients.Clients.Count == 0)
        {
            await resolvedDynamicClients.DisposeAsync();
            return new GatewayResolvedInterceptorClients(_staticClients);
        }

        var combined = new List<McpClient>(_staticClients.Count + resolvedDynamicClients.Clients.Count);
        combined.AddRange(_staticClients);
        combined.AddRange(resolvedDynamicClients.Clients);
        return new GatewayResolvedInterceptorClients(combined, [resolvedDynamicClients]);
    }

    public async ValueTask DisposeAsync()
    {
        if (_clientPool is not null)
        {
            await _clientPool.DisposeAsync();
        }
    }
}

File: typescript/sdk/src/gateway/gateway-interceptor-client-provider.ts
// Copyright 2025 The MCP Interceptors Authors. All rights reserved.
import type { Client } from "@modelcontextprotocol/client";

import type { GatewayMessageContext } from './gateway-message-context.js';
import type { McpInterceptorServerConnectionOptions } from './mcp-interceptor-server-connection-options.js';
import { GatewayInterceptorClientPool } from './gateway-interceptor-client-pool.js';
import { GatewayResolvedInterceptorClients } from './gateway-resolved-interceptor-clients.js';

export type InterceptorServerConnectionResolver = (
  context: GatewayMessageContext,
  event: string,
  signal?: AbortSignal,
) => Promise<McpInterceptorServerConnectionOptions[]>;

export class GatewayInterceptorClientProvider {
  private readonly staticClients: Client[];
  private readonly connectionResolver?: InterceptorServerConnectionResolver;
  private readonly clientPool?: GatewayInterceptorClientPool;

  constructor(staticClients: Client[], connectionResolver?: InterceptorServerConnectionResolver) {
    this.staticClients = staticClients;
    this.connectionResolver = connectionResolver;
    this.clientPool = connectionResolver ? new GatewayInterceptorClientPool() : undefined;
  }

  async resolve(
    messageContext: GatewayMessageContext,
    event: string,
    signal?: AbortSignal,
  ): Promise<GatewayResolvedInterceptorClients> {
    if (!this.connectionResolver) {
      return new GatewayResolvedInterceptorClients(this.staticClients);
    }

    const resolvedConnections = (await this.connectionResolver(messageContext, event, signal)) ?? [];
    const resolvedDynamic = await this.clientPool!.resolveClients(resolvedConnections, signal);

    if (this.staticClients.length === 0) {
      return resolvedDynamic;
    }

    if (resolvedDynamic.clients.length === 0) {
      await resolvedDynamic.dispose();
      return new GatewayResolvedInterceptorClients(this.staticClients);
    }

    return new GatewayResolvedInterceptorClients(
      [...this.staticClients, ...resolvedDynamic.clients],
      [resolvedDynamic],
    );
  }

  async dispose(): Promise<void> {
    if (this.clientPool) {
      await this.clientPool.dispose();
    }
  }
}
